messaging-activemq/src/main/asciidoc/specs/WFLY-5838_artemis_thread_pools.adoc

```diff
@@ -0,0 +1,440 @@
+* keepAlive = 60 seconds
+* workQueue = new SynchronousQueue<Runnable>()
+* threadFactory (groupName = "ActiveMQ-client-global-threads", daemon = true)
+----
+
+== Artemis Global Client Scheduled Thread Pool
+
+By default Artemis configures its client scheduled thread pool at https://github.com/apache/activemq-artemis/blob/master/artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/client/ActiveMQClient.java#L228[ActiveMQClient#getGlobalScheduledThreadPool].
+
+It uses a plain `ScheduledThreadPoolExecutor` with the parameters:
+
+----
+* corePoolSize = 5
+* threadFactory (groupName = "ActiveMQ-client-global-scheduled-threads", daemon = true)
+----
+
+In WildFly's `messaging-activemq` subsystem, the `subsystem` resource provides the `global-client-scheduled-thread-pool-max-size` management attribute to configure the maximum size of this pool. It has no default value   so the default behaviour is to use the valued from Artemis (i.e. `5`)
+
+= Artemis Requirements
+
+Artemis allows the use of external thread pools through its https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/ServiceRegistry.java[ServiceRegistry] interface.
+The WildFly `messaging-activemq` subsystem can then create instance of `ActiveMQServer` and use its `ServiceRegistry` to set the various server thread pools before calling `ActiveMQServer#start` method.
+
+Relevant methods are:
+
+----
+* ServiceRegistry#setExecutorService(ExecutorService)
+* ServiceRegistry#setIOExecutorService(ExecutorService)
+* ServiceRegistry#setScheduledExecutorService(ScheduledExecutorService)
+----
+
+For global thread pools, the  WildFly `messaging-activemq` subsystem must call https://github.com/apache/activemq-artemis/blob/master/artemis-core-client/src/main/java/org/apache/activemq/artemis/api/core/client/ActiveMQClient.java#L197[ActiveMQClient#injectPools] __before any Artemis client or server is started__:
+
+----
+ActiveMQClient#injectPools(ExecutorService, ScheduledExecutorService)
+----
+
+WARNING: Artemis requires that *both the regular and scheduled thread pools are injected* (it is not possible to inject only one and not the other). This must be enforced by WildFly `messaging-activemq` subsystem model.
+
+== Uninjectable Thread Pools in Artemis
+
+There are still a few places where Artemis creates thread pools and do not allow to inject them from WildFly.
+
+* https://github.com/apache/activemq-artemis/blob/master/artemis-server/src/main/java/org/apache/activemq/artemis/core/remoting/impl/invm/InVMConnector.java#L109[InVMConnector#getInVMExecutor]
+** At first glance this ExecutorService should be replaced by the actual global client thread pool returned by `ActiveMQClient#getGlobalThreadPool()` (which is injectable). If the Artemis development team objects that each InVMConnector should have its own unique thread pool, they will have to provide a way to inject it.
```

Comment:
wouldn't that create too many threads?


We would need to create JIRAs and follow up with what is needed.. to a better tracking of what needs to be done.

Reply:
Please note, that separate thread pool for InVmConnector was created as a fix for https://issues.jboss.org/browse/JBEAP-4203